src/platforms/common/px4_work_queue/WorkQueueManager.cpp

```diff
@@ -0,0 +1,315 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "WorkQueueManager.hpp"
+
+#include "WorkQueue.hpp"
+
+#include <string.h>
+
+#include <px4_posix.h>
+#include <px4_tasks.h>
+#include <px4_time.h>
+#include <drivers/drv_hrt.h>
+
+#include <lib/drivers/device/Device.hpp>
+
+using namespace time_literals;
+
+namespace px4
+{
+
+List<WorkQueue *> px4_work_queues_list;
+pthread_mutex_t px4_work_queues_list_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+BlockingQueue<wq_config *, 1> px4_wq_manager_new;
+
+static WorkQueue *
+find_work_queue(const char *name)
+{
+	pthread_mutex_lock(&px4_work_queues_list_mutex);
```

Comment:
Could we do all these things with `std::lock_guard` or is that not available?

Reply:
This is the standard library we currently have available to us. https://github.com/PX4/NuttX/tree/px4_firmware_nuttx-7.28%2B/include/cxx

Reply:
Ah ok that answers my question, so this needs to work on NuttX as well.

Reply:
Here's an option for a proper c++ standard library on NuttX.
https://github.com/PX4/Firmware/pull/7874